### ISSUE TITLE
viewer.close in gui_qt context manager not closing

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#2029]
## 🐛 Bug

I'm running into problems with `viewer.close` not closing inside a `gui_qt` context manager

## To Reproduce

```python
from skimage import data
import napari


with napari.gui_qt():
    # create the viewer with an image
    viewer = napari.view_image(data.astronaut(), rgb=True)
    viewer.close()
```

Leads to me just seeing this

<img width="1200" alt="Screen Shot 2020-12-19 at 12 53 35 PM" src="https://user-images.githubusercontent.com/6531703/102699415-57c32100-41f9-11eb-806f-e5854367957e.png">

and I have to force quit before I can get back control.

## Expected behavior

I would have liked the viewer to close and the script to have finished running and exited

## Environment

This is on latest master, my mac.

**Comment by sofroniewn** [comment]
I'll add that my use case for this is with `napari-animation` where I want to run scripts that save movies and close without having to interact with the viewer. Would be great to have this supported for that functionality!

**Comment by sofroniewn** [comment]
Update - thanks to some tips from @jni I was able to get this working with

```python
from qtpy.QtWidgets import QApplication
app = QApplication([])
```

before making the viewer and no `gui_qt`. Would still be good if we fix the above, but I have what I need for my scripts

**Comment by sofroniewn** [comment]
@tlambert03 my original need for this has gone down (see comments above), but curious what you make of this/ if there is a quick fix for `0.4.3` or if this needs a longer discussion/ more thought

**Comment by tlambert03** [comment]
ah, i missed this one.  Yeah i think the context manager is probably not the correct approach for this.  Remember, `app.exec_()` (and therefore the event loop that would show the viewer) doesn't even run until the context is _exited_ so basically, you've already closed the viewer by the time the app has even started.  The fact that you see anything at all is "lucky". (when I try it I see nothing at all)
If you want to show a viewer for a certain period of time, and have it close all on it's own, you're going to have to hook up events and/or QTimers to the `viewer.close()` method, or to the `app.quit` method.

maybe you can show me a longer example of what you're trying to achieve I can help out.

lastly: 

> and I have to force quit before I can get back control.

often in these cases, you'll have an "empty" napari icon in your dock.  you just need to send the regular mac quit command to it... so I tend to command-tab over to the napari icon, then _keep holding_ the command button and then do command-Q.  usually that will close it and you'll get your terminal back

**Comment by sofroniewn** [comment]
Ok yeah, makes sense to avoid the context manager here. We can pick this up on the `napari-animation` repo with examples there. I'll leave this open for now, but doesn't seem like a priority for `0.4.3` and will bump

**Comment by tlambert03** [comment]
linking to an image.sc post where a similar question was asked: https://forum.image.sc/t/closing-napari-viewer-from-within-python-script/37809

are you leaving this open because you hope it will work someday?  I'm not sure this is an approach that will ever be advisable...

**Comment by sofroniewn** [comment]
Thanks for the link!

> are you leaving this open because you hope it will work someday? I'm not sure this is an approach that will ever be advisable...

I guess I did have a little hope that we might be able to do better. Calling `viewer.close` seems pretty natural, and right now both workaround require direct imports from `qtpy` which seems unfortunate, but even if we don't make any changes we should at least add this to our documentation somewhere. I've added a documentation label now for the issue so we don't lose track of that.

**Comment by tlambert03** [comment]
ok, well let's look at some more real-world examples together and see what we can do

**Comment by tlambert03** [comment]
> both workaround require direct imports from qtpy

well ... if you're doing something with animations, it would be something like `animation.events.finished.connect(viewer.close)` not a direct import from qtpy.  just _something_ that knows when to tell the viewer to close.  I just can't think of many cases in which calling viewer.close procedurally in the context manager is going to work

**Comment by sofroniewn** [comment]
> animation.events.finished.connect(viewer.close)

ooo yeah that's a nice idea! Agreed, this will be best discussed more over some specific examples. I consider it fairly low priority too

**Comment by sofroniewn** [comment]
This was closed by #2056. You no longer need to use a `napari.gui_qt` context manager
